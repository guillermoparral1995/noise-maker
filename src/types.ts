import type { ActionTypes as EnvelopeActionTypes } from './components/Controls/EnvelopeControls/store/actions';
import type { ActionTypes as GeneralControlsActionTypes } from './components/Controls/GeneralControls/store/actions';
import type { ActionTypes as FilterActionTypes } from './components/Controls/FilterControls/store/actions';
import type { ActionTypes as LFOActionTypes } from './components/Controls/LFOControls/store/actions';
import { LFOTargetAction } from './providers/AudioContextProvider';

export type ActionTypes =
  | EnvelopeActionTypes
  | GeneralControlsActionTypes
  | FilterActionTypes
  | LFOTargetAction
  | LFOActionTypes;

export enum Waveform {
  SINE = 'sine',
  SQUARE = 'square',
  SAWTOOTH = 'sawtooth',
  TRIANGLE = 'triangle',
}

export enum FilterType {
  LOWPASS = 'lowpass',
  HIGHPASS = 'highpass',
}

export enum Selectors {
  WAVEFORM = 'waveform',
  FILTER = 'filter',
  LFO_TARGET = 'lfoTarget',
  LFO_WAVEFORM = 'lfoWaveform',
}

export enum Knobs {
  VOLUME = 'VOLUME',
  PAN = 'PAN',
  DETUNE = 'DETUNE',
  ATTACK = 'ATTACK',
  DECAY = 'DECAY',
  SUSTAIN = 'SUSTAIN',
  RELEASE = 'RELEASE',
  FILTER_CUTOFF = 'FILTER_CUTOFF',
  FILTER_RESONANCE = 'FILTER_RESONANCE',
  LFO_FREQUENCY = 'LFO_FREQUENCY',
  LFO_AMPLITUDE = 'LFO_AMPLITUDE',
}

export type LFOTarget =
  | Exclude<
      Knobs,
      | Knobs.LFO_FREQUENCY
      | Knobs.LFO_AMPLITUDE
      | Knobs.ATTACK
      | Knobs.DECAY
      | Knobs.SUSTAIN
      | Knobs.RELEASE
    >
  | 'off';

export interface State {
  volume: number;
  pan: number;
  waveform: Waveform;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  filter: {
    type: FilterType;
    cutoff: number;
    resonance: number;
  };
  lfo: {
    target: LFOTarget;
    waveform: Waveform;
    frequency: number;
    amplitude: number;
  };
}

export enum Actions {
  UPDATE_VOLUME,
  UPDATE_PAN,
  UPDATE_WAVEFORM,
  UPDATE_DETUNE,
  UPDATE_ATTACK,
  UPDATE_DECAY,
  UPDATE_SUSTAIN,
  UPDATE_RELEASE,
  UPDATE_FILTER_TYPE,
  UPDATE_FILTER_CUTOFF,
  UPDATE_FILTER_RESONANCE,
  UPDATE_LFO_TARGET,
  UPDATE_LFO_WAVEFORM,
  UPDATE_LFO_FREQUENCY,
  UPDATE_LFO_AMPLITUDE,
}
export type ActionBuilder<T> = (payload: T) => ActionTypes;
