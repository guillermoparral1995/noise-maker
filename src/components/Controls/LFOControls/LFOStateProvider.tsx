import React, { PropsWithChildren, useReducer } from 'react';
import initialState, { LFOControlsState } from './store/initialState';
import { ActionTypes } from './store/actions';
import reducer from './store/reducer';

interface LFOStateProviderValue {
  state: LFOControlsState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const lfoStateContext =
  React.createContext<LFOStateProviderValue>(undefined);

export const LFOStateProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <lfoStateContext.Provider value={{ state, dispatch }}>
      {children}
    </lfoStateContext.Provider>
  );
};
