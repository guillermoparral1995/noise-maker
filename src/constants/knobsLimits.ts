import { Knobs } from '../types';

interface KnobLimit {
  min: number;
  max: number;
  default: number;
}

export const knobsLimits: Record<Knobs, KnobLimit> = {
  [Knobs.VOLUME]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
  [Knobs.PAN]: {
    min: -1,
    max: 1,
    default: 0,
  },
  [Knobs.ATTACK]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.DECAY]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.SUSTAIN]: {
    min: 0,
    max: 1,
    default: 1,
  },
  [Knobs.RELEASE]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.FILTER_CUTOFF]: {
    min: 10,
    max: 5000,
    default: 350,
  },
  [Knobs.FILTER_RESONANCE]: {
    min: 0.0001,
    max: 50,
    default: 1,
  },
  [Knobs.LFO_FREQUENCY]: {
    min: 0.2,
    max: 10,
    default: 1,
  },
  [Knobs.LFO_AMPLITUDE]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
};
