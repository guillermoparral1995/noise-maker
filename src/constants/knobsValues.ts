import {
  updateAttack,
  updateDecay,
  updateDetune,
  updatePitchbend,
  updateRelease,
  updateSustain,
} from '../components/Controls/EnvelopeControls/store/actions';
import {
  updateFilterCutoff,
  updateFilterResonance,
} from '../components/Controls/FilterControls/store/actions';
import {
  updatePan,
  updateVolume,
} from '../components/Controls/GeneralControls/store/actions';
import {
  updateLFO1Amplitude,
  updateLFO1Frequency,
  updateLFO2Amplitude,
  updateLFO2Frequency,
} from '../components/Controls/LFOControls/store/actions';
import { ActionBuilder, Knobs } from '../types';

interface KnobValue {
  label: string;
  min: number;
  max: number;
  default: number;
  midiControl?: number;
  action: ActionBuilder<number>;
  step?: number;
}

export const knobsValues: Record<Knobs, KnobValue> = {
  [Knobs.VOLUME]: {
    label: 'Volume',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 74,
    action: updateVolume,
  },
  [Knobs.PAN]: {
    label: 'Pan',
    min: -1,
    max: 1,
    default: 0,
    midiControl: 18,
    action: updatePan,
  },
  [Knobs.DETUNE]: {
    label: 'Detune',
    min: -1200,
    max: 1200,
    default: 0,
    midiControl: 75,
    action: updateDetune,
    step: 1,
  },
  [Knobs.PITCHBEND]: {
    label: 'Pitchbend',
    min: -300,
    max: 300,
    default: 0,
    action: updatePitchbend,
    step: 1,
  },
  [Knobs.ATTACK]: {
    label: 'Attack',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 93,
    action: updateAttack,
  },
  [Knobs.DECAY]: {
    label: 'Decay',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 91,
    action: updateDecay,
  },
  [Knobs.SUSTAIN]: {
    label: 'Sustain',
    min: 0,
    max: 1,
    default: 1,
    midiControl: 73,
    action: updateSustain,
  },
  [Knobs.RELEASE]: {
    label: 'Release',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 79,
    action: updateRelease,
  },
  [Knobs.FILTER_CUTOFF]: {
    label: 'Cut-off',
    min: 10,
    max: 5000,
    default: 350,
    midiControl: 71,
    action: updateFilterCutoff,
    step: 1,
  },
  [Knobs.FILTER_RESONANCE]: {
    label: 'Resonance',
    min: 0.0001,
    max: 50,
    default: 1,
    midiControl: 19,
    action: updateFilterResonance,
    step: 0.1,
  },
  [Knobs.LFO_1_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 76,
    action: updateLFO1Frequency,
  },
  [Knobs.LFO_1_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 16,
    action: updateLFO1Amplitude,
  },
  [Knobs.LFO_2_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 77,
    action: updateLFO2Frequency,
  },
  [Knobs.LFO_2_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 17,
    action: updateLFO2Amplitude,
  },
};
