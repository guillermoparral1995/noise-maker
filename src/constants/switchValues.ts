import { switchDelayTrails } from '../components/Controls/DelayControls/store/actions';
import { switchDelay } from '../providers/AudioContextProvider/store/actions';
import { Switchs } from '../types';

export const switchValues = {
  [Switchs.DELAY]: {
    label: 'Delay',
    action: switchDelay,
  },

  [Switchs.DELAY_TRAILS]: {
    label: 'Trails',
    action: switchDelayTrails,
  },
};
