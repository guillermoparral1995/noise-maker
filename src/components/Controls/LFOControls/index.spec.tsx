import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { LFOControls_ as LFOControls } from '.';
import {
  AudioContextMock,
  GainNodeMock,
  OscillatorNodeMock,
} from '../../../../__mocks__';
import { knobsValues } from '../../../constants/knobsValues';
import {
  AudioContextProvider,
  LFOMock,
} from '../../../providers/AudioContextProvider';
import {
  withMockedMIDIInput,
  withMockedMIDINoInput,
} from '../../../providers/MIDIProvider';
import { Actions, Knobs, LFO1Target, LFO2Target } from '../../../types';
import { LFOStateProvider } from './LFOStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

describe('LFO controls', () => {
  const mockContext = new AudioContextMock();
  const mockLfo1: LFOMock<LFO1Target> = {
    output: new GainNodeMock(),
  };
  const mockLfo2: LFOMock<LFO2Target> = {
    output: new GainNodeMock(),
  };

  afterEach(() => {
    cleanup();
  });

  it('should render correctly', async () => {
    const mockLfo1Oscillator = new OscillatorNodeMock();
    const mockLfo2Oscillator = new OscillatorNodeMock();
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider>
            <div data-testid="lfo-controls">
              <LFOControls
                __mockLFO1={mockLfo1Oscillator}
                __mockLFO2={mockLfo2Oscillator}
              ></LFOControls>
            </div>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );

    const controls = await findByTestId('lfo-controls');
    expect(mockLfo1Oscillator.__mockConnect).toHaveBeenCalledWith(
      mockLfo1.output,
    );
    expect(mockLfo2Oscillator.__mockConnect).toHaveBeenCalledWith(
      mockLfo2.output,
    );
    expect(mockLfo1Oscillator.__mockStart).toHaveBeenCalled();
    expect(mockLfo2Oscillator.__mockStart).toHaveBeenCalled();
    expect(controls).toMatchSnapshot();
  });

  it('should update lfo 1 output gain when target changes ', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider>
            <LFOControls></LFOControls>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );

    await findByTestId(Knobs.LFO_1_FREQUENCY);
    await userEvent.click(screen.getAllByRole('button')[0]);
    await userEvent.click(screen.getByText('Volume'));

    expect(mockLfo1.output.gain.value).toEqual(0.25);
  });

  it('should update lfo 2 output gain when target changes ', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider>
            <LFOControls></LFOControls>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );

    await findByTestId(Knobs.LFO_1_FREQUENCY);
    await userEvent.click(screen.getAllByRole('button')[2]);

    await userEvent.click(screen.getByText('LFO 1 Frequency'));

    expect(mockLfo2.output.gain.value).toEqual(2.45);
  });

  it('should trigger lfo 1 dispatch on MIDI knob change', async () => {
    const mockDispatch = jest.fn();
    let mockEventName: string;
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        mockEventName = eventName;
        listener({
          value: 1,
          controller: {
            number: knobsValues[Knobs.LFO_1_FREQUENCY].midiControl,
          },
        });
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider __mockDispatch={mockDispatch}>
            <LFOControls></LFOControls>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );

    await findByTestId(Knobs.LFO_1_FREQUENCY);
    expect(mockEventName).toEqual('controlchange');
    expect(mockDispatch).toHaveBeenCalledWith({
      type: Actions.UPDATE_LFO_1_FREQUENCY,
      payload: 10,
    });
  });

  it('should trigger lfo 2 dispatch on MIDI knob change', async () => {
    const mockDispatch = jest.fn();
    let mockEventName: string;
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        mockEventName = eventName;
        listener({
          value: 0.1,
          controller: {
            number: knobsValues[Knobs.LFO_2_AMPLITUDE].midiControl,
          },
        });
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider __mockDispatch={mockDispatch}>
            <LFOControls></LFOControls>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );

    await findByTestId(Knobs.LFO_2_AMPLITUDE);
    expect(mockEventName).toEqual('controlchange');
    expect(mockDispatch).toHaveBeenCalledWith({
      type: Actions.UPDATE_LFO_2_AMPLITUDE,
      payload: 0.1,
    });
  });

  it('should connect lfo2 with param', async () => {
    const mockLfo2: LFOMock<LFO2Target> = {
      target: Knobs.LFO_1_AMPLITUDE,
      output: new GainNodeMock(),
    };
    const mockDispatch = jest.fn();
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <LFOStateProvider __mockDispatch={mockDispatch}>
            <LFOControls></LFOControls>
          </LFOStateProvider>
        </AudioContextProvider>,
      ),
    );
    await findByTestId(Knobs.LFO_2_AMPLITUDE);

    expect(mockLfo2.output.__mockConnect).toHaveBeenCalledWith(
      mockLfo1.output.gain,
    );
  });
});
