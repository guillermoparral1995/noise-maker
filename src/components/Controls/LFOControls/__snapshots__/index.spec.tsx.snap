// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LFO controls should render correctly 1`] = `
<div
  data-testid="lfo-controls"
>
  <div
    class="lfo_column"
  >
    <h3>
      LFO 1
    </h3>
    <span
      class="p-float-label selector"
      data-testid="LFO_1_TARGET"
    >
      <div
        class="p-dropdown p-component p-inputwrapper p-inputwrapper-filled"
        data-p-disabled="false"
        data-p-focus="false"
        data-pc-name="dropdown"
        data-pc-section="root"
      >
        <div
          class="p-hidden-accessible"
          data-pc-section="hiddenselectedmessage"
        >
          <input
            aria-haspopup="listbox"
            data-pc-section="input"
            id="LFO_1_TARGET"
            readonly=""
            type="text"
          />
        </div>
        <div
          class="p-hidden-accessible p-dropdown-hidden-select"
          data-pc-section="hiddenselectedmessage"
        >
          <select
            aria-hidden="true"
            data-pc-section="select"
            name="LFO_1_TARGET"
            tabindex="-1"
          >
            <option
              data-pc-section="option"
              selected=""
              value="off"
            >
              Off
            </option>
          </select>
        </div>
        <span
          class="p-dropdown-label p-inputtext"
          data-pc-section="input"
        >
          Off
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="p-dropdown-trigger"
          data-pc-section="trigger"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="p-icon p-dropdown-trigger-icon p-clickable"
            data-pc-section="dropdownicon"
            data-pr-overlay-visible="false"
            fill="none"
            height="14"
            viewBox="0 0 14 14"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.01744 10.398C6.91269 10.3985 6.8089 10.378 6.71215 10.3379C6.61541 10.2977 6.52766 10.2386 6.45405 10.1641L1.13907 4.84913C1.03306 4.69404 0.985221 4.5065 1.00399 4.31958C1.02276 4.13266 1.10693 3.95838 1.24166 3.82747C1.37639 3.69655 1.55301 3.61742 1.74039 3.60402C1.92777 3.59062 2.11386 3.64382 2.26584 3.75424L7.01744 8.47394L11.769 3.75424C11.9189 3.65709 12.097 3.61306 12.2748 3.62921C12.4527 3.64535 12.6199 3.72073 12.7498 3.84328C12.8797 3.96582 12.9647 4.12842 12.9912 4.30502C13.0177 4.48162 12.9841 4.662 12.8958 4.81724L7.58083 10.1322C7.50996 10.2125 7.42344 10.2775 7.32656 10.3232C7.22968 10.3689 7.12449 10.3944 7.01744 10.398Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <label
        for="LFO_1_TARGET"
      >
        Target
      </label>
    </span>
    <span
      class="p-float-label selector"
      data-testid="LFO_1_WAVEFORM"
    >
      <div
        class="p-dropdown p-component p-inputwrapper p-inputwrapper-filled"
        data-p-disabled="false"
        data-p-focus="false"
        data-pc-name="dropdown"
        data-pc-section="root"
      >
        <div
          class="p-hidden-accessible"
          data-pc-section="hiddenselectedmessage"
        >
          <input
            aria-haspopup="listbox"
            data-pc-section="input"
            id="LFO_1_WAVEFORM"
            readonly=""
            type="text"
          />
        </div>
        <div
          class="p-hidden-accessible p-dropdown-hidden-select"
          data-pc-section="hiddenselectedmessage"
        >
          <select
            aria-hidden="true"
            data-pc-section="select"
            name="LFO_1_WAVEFORM"
            tabindex="-1"
          >
            <option
              data-pc-section="option"
              selected=""
              value="sine"
            >
              Sine
            </option>
          </select>
        </div>
        <span
          class="p-dropdown-label p-inputtext"
          data-pc-section="input"
        >
          Sine
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="p-dropdown-trigger"
          data-pc-section="trigger"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="p-icon p-dropdown-trigger-icon p-clickable"
            data-pc-section="dropdownicon"
            data-pr-overlay-visible="false"
            fill="none"
            height="14"
            viewBox="0 0 14 14"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.01744 10.398C6.91269 10.3985 6.8089 10.378 6.71215 10.3379C6.61541 10.2977 6.52766 10.2386 6.45405 10.1641L1.13907 4.84913C1.03306 4.69404 0.985221 4.5065 1.00399 4.31958C1.02276 4.13266 1.10693 3.95838 1.24166 3.82747C1.37639 3.69655 1.55301 3.61742 1.74039 3.60402C1.92777 3.59062 2.11386 3.64382 2.26584 3.75424L7.01744 8.47394L11.769 3.75424C11.9189 3.65709 12.097 3.61306 12.2748 3.62921C12.4527 3.64535 12.6199 3.72073 12.7498 3.84328C12.8797 3.96582 12.9647 4.12842 12.9912 4.30502C13.0177 4.48162 12.9841 4.662 12.8958 4.81724L7.58083 10.1322C7.50996 10.2125 7.42344 10.2775 7.32656 10.3232C7.22968 10.3689 7.12449 10.3944 7.01744 10.398Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <label
        for="LFO_1_WAVEFORM"
      >
        Waveform
      </label>
    </span>
    <div
      data-testid="LFO_1_FREQUENCY"
      tabindex="0"
    >
      <label
        class="label"
        for="LFO_1_FREQUENCY"
      >
        Frequency
      </label>
      <div
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="LFO_1_FREQUENCY"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 17.4395165897366 73.2339174503561"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="LFO_1_FREQUENCY"
            text-anchor="middle"
            x="50"
            y="57"
          >
            1
          </text>
        </svg>
      </div>
    </div>
    <div
      data-testid="LFO_1_AMPLITUDE"
      tabindex="0"
    >
      <label
        class="label"
        for="LFO_1_AMPLITUDE"
      >
        Amplitude
      </label>
      <div
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="LFO_1_AMPLITUDE"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 50 10"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="LFO_1_AMPLITUDE"
            text-anchor="middle"
            x="50"
            y="57"
          >
            0.5
          </text>
        </svg>
      </div>
    </div>
  </div>
  <div
    class="lfo_column"
  >
    <h3>
      LFO 2
    </h3>
    <span
      class="p-float-label selector"
      data-testid="LFO_2_TARGET"
    >
      <div
        class="p-dropdown p-component p-inputwrapper p-inputwrapper-filled"
        data-p-disabled="false"
        data-p-focus="false"
        data-pc-name="dropdown"
        data-pc-section="root"
      >
        <div
          class="p-hidden-accessible"
          data-pc-section="hiddenselectedmessage"
        >
          <input
            aria-haspopup="listbox"
            data-pc-section="input"
            id="LFO_2_TARGET"
            readonly=""
            type="text"
          />
        </div>
        <div
          class="p-hidden-accessible p-dropdown-hidden-select"
          data-pc-section="hiddenselectedmessage"
        >
          <select
            aria-hidden="true"
            data-pc-section="select"
            name="LFO_2_TARGET"
            tabindex="-1"
          >
            <option
              data-pc-section="option"
              selected=""
              value="off"
            >
              Off
            </option>
          </select>
        </div>
        <span
          class="p-dropdown-label p-inputtext"
          data-pc-section="input"
        >
          Off
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="p-dropdown-trigger"
          data-pc-section="trigger"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="p-icon p-dropdown-trigger-icon p-clickable"
            data-pc-section="dropdownicon"
            data-pr-overlay-visible="false"
            fill="none"
            height="14"
            viewBox="0 0 14 14"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.01744 10.398C6.91269 10.3985 6.8089 10.378 6.71215 10.3379C6.61541 10.2977 6.52766 10.2386 6.45405 10.1641L1.13907 4.84913C1.03306 4.69404 0.985221 4.5065 1.00399 4.31958C1.02276 4.13266 1.10693 3.95838 1.24166 3.82747C1.37639 3.69655 1.55301 3.61742 1.74039 3.60402C1.92777 3.59062 2.11386 3.64382 2.26584 3.75424L7.01744 8.47394L11.769 3.75424C11.9189 3.65709 12.097 3.61306 12.2748 3.62921C12.4527 3.64535 12.6199 3.72073 12.7498 3.84328C12.8797 3.96582 12.9647 4.12842 12.9912 4.30502C13.0177 4.48162 12.9841 4.662 12.8958 4.81724L7.58083 10.1322C7.50996 10.2125 7.42344 10.2775 7.32656 10.3232C7.22968 10.3689 7.12449 10.3944 7.01744 10.398Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <label
        for="LFO_2_TARGET"
      >
        Target
      </label>
    </span>
    <span
      class="p-float-label selector"
      data-testid="LFO_2_WAVEFORM"
    >
      <div
        class="p-dropdown p-component p-inputwrapper p-inputwrapper-filled"
        data-p-disabled="false"
        data-p-focus="false"
        data-pc-name="dropdown"
        data-pc-section="root"
      >
        <div
          class="p-hidden-accessible"
          data-pc-section="hiddenselectedmessage"
        >
          <input
            aria-haspopup="listbox"
            data-pc-section="input"
            id="LFO_2_WAVEFORM"
            readonly=""
            type="text"
          />
        </div>
        <div
          class="p-hidden-accessible p-dropdown-hidden-select"
          data-pc-section="hiddenselectedmessage"
        >
          <select
            aria-hidden="true"
            data-pc-section="select"
            name="LFO_2_WAVEFORM"
            tabindex="-1"
          >
            <option
              data-pc-section="option"
              selected=""
              value="sine"
            >
              Sine
            </option>
          </select>
        </div>
        <span
          class="p-dropdown-label p-inputtext"
          data-pc-section="input"
        >
          Sine
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="p-dropdown-trigger"
          data-pc-section="trigger"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="p-icon p-dropdown-trigger-icon p-clickable"
            data-pc-section="dropdownicon"
            data-pr-overlay-visible="false"
            fill="none"
            height="14"
            viewBox="0 0 14 14"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.01744 10.398C6.91269 10.3985 6.8089 10.378 6.71215 10.3379C6.61541 10.2977 6.52766 10.2386 6.45405 10.1641L1.13907 4.84913C1.03306 4.69404 0.985221 4.5065 1.00399 4.31958C1.02276 4.13266 1.10693 3.95838 1.24166 3.82747C1.37639 3.69655 1.55301 3.61742 1.74039 3.60402C1.92777 3.59062 2.11386 3.64382 2.26584 3.75424L7.01744 8.47394L11.769 3.75424C11.9189 3.65709 12.097 3.61306 12.2748 3.62921C12.4527 3.64535 12.6199 3.72073 12.7498 3.84328C12.8797 3.96582 12.9647 4.12842 12.9912 4.30502C13.0177 4.48162 12.9841 4.662 12.8958 4.81724L7.58083 10.1322C7.50996 10.2125 7.42344 10.2775 7.32656 10.3232C7.22968 10.3689 7.12449 10.3944 7.01744 10.398Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <label
        for="LFO_2_WAVEFORM"
      >
        Waveform
      </label>
    </span>
    <div
      data-testid="LFO_2_FREQUENCY"
      tabindex="0"
    >
      <label
        class="label"
        for="LFO_2_FREQUENCY"
      >
        Frequency
      </label>
      <div
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="LFO_2_FREQUENCY"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 17.4395165897366 73.2339174503561"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="LFO_2_FREQUENCY"
            text-anchor="middle"
            x="50"
            y="57"
          >
            1
          </text>
        </svg>
      </div>
    </div>
    <div
      data-testid="LFO_2_AMPLITUDE"
      tabindex="0"
    >
      <label
        class="label"
        for="LFO_2_AMPLITUDE"
      >
        Amplitude
      </label>
      <div
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="LFO_2_AMPLITUDE"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 50 10"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="LFO_2_AMPLITUDE"
            text-anchor="middle"
            x="50"
            y="57"
          >
            0.5
          </text>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
